<a href="https://colab.research.google.com/github/twyeh/AI-in-education/blob/main/%E5%96%AE%E5%B1%A4%E7%A5%9E%E7%B6%93%E5%85%83_XOR%E9%96%98%2B_%E6%90%8D%E5%A4%B1%E8%A8%93%E7%B7%B4_%E7%AF%84%E6%9C%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**目標**

1. 設計一個一位元XOR閘模擬類神經元網路
2. 透過計算學習損失函數進行學習訓練

步驟 1：導入必要的函式庫

In [8]:
import numpy as np

步驟 2：定義 XOR 閘的資料集

In [9]:
# 輸入
X = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
# 輸出
y = np.array([0, 1, 1, 0]) # XOR 目標
#y = np.array([0, 0, 0, 1]) # AND 目標
#y = np.array([0, 1, 1, 1]) # OR 目標

步驟 3：定義類神經元網路的架構

In [10]:
# 定義具有兩個輸入神經元、兩個隱藏神經元和一個輸出神經元的類神經元網路
# 輸入層：2 個神經元
# 隱藏層：2 個神經元，使用 sigmoid 激活函數
# 輸出層：1 個神經元，使用 sigmoid 激活函數
def sigmoid(x):
  return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
  return x * (1 - x)

步驟 4：初始化類神經元網路的權重和偏差

In [11]:
# 初始化權重和偏差為隨機值
# 隱藏層權重
hidden_weights = np.random.uniform(size=(2, 2))
# 隱藏層偏差
hidden_bias = np.random.uniform(size=(1, 2))
# 輸出層權重
output_weights = np.random.uniform(size=(2, 1))
# 輸出層偏差
output_bias = np.random.uniform(size=(1, 1))

步驟 5：訓練類神經元網路

In [15]:
# 設定學習率和迭代次數
learning_rate = 0.1
epochs = 10000

# 訓練迴圈
for epoch in range(epochs):
  # 正向傳播
  hidden_layer_activation = np.dot(X, hidden_weights) + hidden_bias
  hidden_layer_output = sigmoid(hidden_layer_activation)
  output_layer_activation = np.dot(hidden_layer_output, output_weights) + output_bias
  predicted_output = sigmoid(output_layer_activation)

  # 反向傳播
  error = y.reshape(-1, 1) - predicted_output
  d_predicted_output = error * sigmoid_derivative(predicted_output)
  error_hidden_layer = d_predicted_output.dot(output_weights.T)
  d_hidden_layer = error_hidden_layer * sigmoid_derivative(hidden_layer_output)

  # 更新權重和偏差
  output_weights += hidden_layer_output.T.dot(d_predicted_output) * learning_rate
  output_bias += np.sum(d_predicted_output, axis=0, keepdims=True) * learning_rate
  hidden_weights += X.T.dot(d_hidden_layer) * learning_rate
  hidden_bias += np.sum(d_hidden_layer, axis=0, keepdims=True) * learning_rate

  # 計算損失函數（均方誤差）
  loss = np.mean(np.square(error))

  # 每 1000 個迭代印出損失
  if epoch % 1000 == 0:
    print(f"Epoch: {epoch}, Loss: {loss}")

Epoch: 0, Loss: 0.0005720896404361145
Epoch: 1000, Loss: 0.000548263241504109
Epoch: 2000, Loss: 0.0005262952376861805
Epoch: 3000, Loss: 0.0005059787653683216
Epoch: 4000, Loss: 0.0004871363621200064
Epoch: 5000, Loss: 0.00046961494575593297
Epoch: 6000, Loss: 0.0004532817850189266
Epoch: 7000, Loss: 0.0004380212412461737
Epoch: 8000, Loss: 0.00042373211464864763
Epoch: 9000, Loss: 0.00041032546855687506


**損失函數**

* 這個程式碼使用均方誤差 (MSE) 作為損失函數。
* MSE 計算預測輸出和實際輸出之間的平均平方差。
* 訓練類神經元網路的目標是最小化損失函數。

學習結果

In [6]:
print(f"hidden weights: {hidden_weights}, bias: {hidden_bias}")
print(f"output weights: {output_weights}, bias: {output_bias}'")

hidden weights: [[3.6653158  5.72633095]
 [3.6798572  5.79868177]], bias: [[-5.62242328 -2.39363783]]
output weights: [[-8.03488326]
 [ 7.40847508]], bias: [[-3.33382908]]'


XOR 閘學習模型：
\begin{eqnarray}
w_h &=&
\begin{cases}
3.66 & 5.73 \\
3.68 & 5.80
\end{cases}
\\
b_h &=& \begin{cases}
-5.52 & -2.39
\end{cases}\\
w_o &=&
\begin{cases}
-8.0. & 7.41  
\end{cases}
\\
b_o &=& \begin{cases}
-3.33
\end{cases}
\end{eqnarray}

步驟 6：評估訓練好的類神經元網路

In [7]:
# 預測 XOR 閘的輸出
hidden_layer_activation = np.dot(X, hidden_weights) + hidden_bias
hidden_layer_output = sigmoid(hidden_layer_activation)
output_layer_activation = np.dot(hidden_layer_output, output_weights) + output_bias
predicted_output = sigmoid(output_layer_activation)

print("預測輸出：")
print(predicted_output)

預測輸出：
[[0.06048372]
 [0.94423602]
 [0.94399821]
 [0.06044827]]


XOR 閘學習預測：
\begin{eqnarray}
Y_O &=&
\begin{cases}
0.06 \\
0.94 \\
0.94 \\
0.06
\end{cases}
\end{eqnarray}